### PR TITLE
support Python 3.7 and 3.8

### PR DESCRIPTION
**Pull request by thrix**
To support these we need to do some major upgrades of test tools
and ruamel.yaml.

I was able to fix all tests, most things were fixed for new flake8, please review carefully, some things are not exactly clear to me and might be problematic.

Trying to run `gluetool-modules` tests against it.

Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

### CODE REVIEW

**Comment by thrix** [comment]
meh few more tweaks needed

**Comment by thrix** [comment]
:rocket: https://travis-ci.org/github/gluetool/gluetool/builds/727509995

**Comment by thrix** [comment]
All tests pass \o/

File: gluetool/action.py [hunk 1]
```diff
@@ -81,11 +81,11 @@
...
 DEFAULT_TRACING_SERVICE_NAME = 'gluetool'
 DEFAULT_TRACING_REPORTING_HOST = '127.0.0.1'
-DEFAULT_TRACING_REPORTING_PORT = 5775
-DEFAULT_TRACING_FLUSH_TIMEOUT = 30
+DEFAULT_TRACING_REPORTING_PORT = '5775'
+DEFAULT_TRACING_FLUSH_TIMEOUT = '30'
 
 
-class Tracer(object):
+class Tracer():
```
Comment (happz) [inline gluetool/action.py]: Why this change? It won't work in Python 2.7 anymore, making `Tracer` an old-style class.

File: gluetool/action.py [hunk 2]
```diff
@@ -81,11 +81,11 @@
 
 DEFAULT_TRACING_SERVICE_NAME = 'gluetool'
 DEFAULT_TRACING_REPORTING_HOST = '127.0.0.1'
-DEFAULT_TRACING_REPORTING_PORT = 5775
-DEFAULT_TRACING_FLUSH_TIMEOUT = 30
+DEFAULT_TRACING_REPORTING_PORT = '5775'
+DEFAULT_TRACING_FLUSH_TIMEOUT = '30'
```
Comment (happz) [inline gluetool/action.py]: These values represent port *number* and timeout, which is in seconds - number again. Why changing them to strings?

File: setup.py [hunk 3]
```diff
@@ -56,10 +56,7 @@
               'raven==6.9.0',
               'requests==2.19.1',
               'requests-toolbelt==0.8.0',
-              # since 0.15.52, it returns ordereddict instead of pure dicts, making format_dict unhappy
-              'ruamel.yaml==0.15.51',
-              # newer versions bring way too many incompabilities, needs deeper inspection
-              # 'ruamel.yaml==0.15.34',
+              'ruamel.yaml==0.16.12',
```
Comment (happz) [inline setup.py:59]: Have you checked what you get when you `load_yaml` and try to print it with `format_dict`? `ruamel.yaml` started to use `OrderedDict` instead of plain mappings, killing any readability. Hence the comment...

File: gluetool_modules/dep_list.py [hunk 4]
```diff
@@ -97,6 +97,7 @@ def pip_version_unify(self, pip_deps):
     def get_allowed_version_bounds(self, item):
         if not item.equal and not item.lower and not item.upper:
             return item.pkg
+        # pylint: disable=no-else-return
```
Comment (happz) [inline gluetool_modules/dep_list.py]: This is actually something we should rather fix than silence.

```python
        if item.equal:
            ...
            return '{}=={}'.format(item.pkg, item.equal)
        elif item.upper and item.lower:
```

would become:

```python
        if item.equal:
            ...
            return '{}=={}'.format(item.pkg, item.equal)
        if item.upper and item.lower:
```

`elif` is not necessary, the previous condition ends with `return`, therefore either the condition was true-ish, then you leave the function through this condition, or it was false-ish, and you can test again with plain `if`.

File: gluetool/action.py (hunk 1 above)
Comment (thrix) [reply gluetool/action.py]: a leftover from experiments :/

File: setup.py (hunk 3 above)
Comment (thrix) [reply setup.py:59]: not yet

File: gluetool_modules/dep_list.py (hunk 4 above)
Comment (thrix) [reply gluetool_modules/dep_list.py]: right, was not getting it yesterday, thanks

Comment (thrix) [reply gluetool_modules/dep_list.py]: fixed

File: setup.py (hunk 3 above)
Comment (thrix) [reply setup.py:59]: @happz hmm, seems good to me?

```python
> from gluetool.utils import load_yaml
> from gluetool.log import format_dict
> load_yaml('generate-docs.yml')
[ordereddict([('hosts', 'localhost'), ('vars', ordereddict([('sphinx_builder', 'html'), ('generate_dir', './docs'), ('log_path', '{{ generate_dir }}/build/sphinx_build_stderr.log'), ('clean', False), ('static_rst', ['source/framework.rst', 'source/modules.rst', 'source/howto-tests.rst', 'source/howto-docs.rst', 'source/index.rst', 'source/DEVELOPMENT.rst']), ('base_files', ['source/conf.py']), ('clean_files', ['build', 'source/gluetool.*'])])), ('tasks', [ordereddict([('name', 'Ensure directory hierarchy exists'), ('file', ordereddict([('path', '{{ item }}'), ('state', 'directory')])), ('with_items', ['{{ generate_dir }}/', '{{ generate_dir }}/source/'])]), ordereddict([('name', "Clean doc in '{{ generate_dir }}'"), ('shell', 'rm -rf {{ generate_dir }}/{{ item }}'), ('args', ordereddict([('removes', '{{ generate_dir }}/{{ item }}')])), ('when', 'clean'), ('with_items', '{{ clean_files }}'), ('tags', ['clean'])]), ordereddict([('name', 'Copy static_rst files'), ('copy', ordereddict([('src', 'docs/{{ item }}'), ('dest', '{{ generate_dir }}/{{ item }}')])), ('with_items', '{{ static_rst }}')]), ordereddict([('name', 'Copy base files'), ('copy', ordereddict([('src', 'docs/{{ item }}'), ('dest', '{{ generate_dir }}/{{ item }}')])), ('with_items', '{{ base_files }}')]), ordereddict([('name', 'Generate apidoc'), ('command', 'sphinx-apidoc -T -e -o "{{ generate_dir }}/source/" gluetool/')]), ordereddict([('name', 'Generate documentation'), ('command', 'sphinx-build -M "{{ sphinx_builder }}" source build'), ('register', 'sphinx_build'), ('args', ordereddict([('chdir', '{{ generate_dir }}')]))]), ordereddict([('name', 'Create error log file {{log_path}}'), ('copy', ordereddict([('content', '{{ sphinx_build.stderr }}'), ('dest', '{{ log_path }}')]))]), ordereddict([('name', 'Check log file for errors'), ('command', 'awk \'/ERROR:/ { print "Errors were detected, check log file and your docs"; exit 1; }\' "{{ log_path }}"')])])])]
> print(format_dict(load_yaml('generate-docs.yml')))
[
    {
        "hosts": "localhost",
        "tasks": [
            {
                "file": {
                    "path": "{{ item }}",
                    "state": "directory"
                },
                "name": "Ensure directory hierarchy exists",
                "with_items": [
                    "{{ generate_dir }}/",
                    "{{ generate_dir }}/source/"
                ]
            },
            {
                "args": {
                    "removes": "{{ generate_dir }}/{{ item }}"
                },
                "name": "Clean doc in '{{ generate_dir }}'",
                "shell": "rm -rf {{ generate_dir }}/{{ item }}",
                "tags": [
                    "clean"
                ],
                "when": "clean",
                "with_items": "{{ clean_files }}"
            },
            {
                "copy": {
                    "dest": "{{ generate_dir }}/{{ item }}",
                    "src": "docs/{{ item }}"
                },
                "name": "Copy static_rst files",
                "with_items": "{{ static_rst }}"
            },
            {
                "copy": {
                    "dest": "{{ generate_dir }}/{{ item }}",
                    "src": "docs/{{ item }}"
                },
                "name": "Copy base files",
                "with_items": "{{ base_files }}"
            },
            {
                "command": "sphinx-apidoc -T -e -o \"{{ generate_dir }}/source/\" gluetool/",
                "name": "Generate apidoc"
            },
            {
                "args": {
                    "chdir": "{{ generate_dir }}"
                },
                "command": "sphinx-build -M \"{{ sphinx_builder }}\" source build",
                "name": "Generate documentation",
                "register": "sphinx_build"
            },
            {
                "copy": {
                    "content": "{{ sphinx_build.stderr }}",
                    "dest": "{{ log_path }}"
                },
                "name": "Create error log file {{log_path}}"
            },
            {
                "command": "awk '/ERROR:/ { print \"Errors were detected, check log file and your docs\"; exit 1; }' \"{{ log_path }}\"",
                "name": "Check log file for errors"
            }
        ],
        "vars": {
            "base_files": [
                "source/conf.py"
            ],
            "clean": false,
            "clean_files": [
                "build",
                "source/gluetool.*"
            ],
            "generate_dir": "./docs",
            "log_path": "{{ generate_dir }}/build/sphinx_build_stderr.log",
            "sphinx_builder": "html",
            "static_rst": [
                "source/framework.rst",
                "source/modules.rst",
                "source/howto-tests.rst",
                "source/howto-docs.rst",
                "source/index.rst",
                "source/DEVELOPMENT.rst"
            ]
        }
    }
]
```

File: gluetool/action.py (hunk 2 above)
Comment (thrix) [reply gluetool/action.py]: @happz yes, there was a pylint error:

https://github.com/PyCQA/pylint/issues/1669

Comment (happz) [reply gluetool/action.py]: Then pylint is wrong. I'm not masking numbers as strings to make some linter happy. It's a number, and Python luckily does have a data type for handling numbers.

File: setup.py (hunk 3 above)
Comment (happz) [reply setup.py:59]: Looks much better than it used too. When I tested it, the output looked like what you got from `load_yaml()`.

File: gluetool/action.py (hunk 2 above)
Comment (thrix) [reply gluetool/action.py]: ack

Comment (thrix) [reply gluetool/action.py]: @happz added ignore for the warning

**Comment by thrix** [comment]
@happz all issues resolved

**Review by ukulekek — DISMISSED** [review]

**Comment by thrix** [comment]
tests are not super stable, I am hitting occasionally hitting errors with `cyclic-import`, after rerun, they pass :( See for example:

https://travis-ci.org/github/gluetool/gluetool/builds/728810496

**Comment by thrix** [comment]
@happz I would disable that test, I cannot reproduce it locally, restarting can fix the problem, not sure what else to do :/

**Comment by thrix** [comment]
@optak @ukulekek @akhait @mkluson  pls review if you have a bit of time

**Review by happz — APPROVED** [review]

**Comment by happz** [comment]
Approved.

I'm really looking forward to removing Pylint completely. It's so slow and so high-maintenance... So many exceptions to get perfectly valid code through :/

**Comment by thrix** [comment]
@optak @ukulekek @akhait @mkluson @Madeyro kindly poking for review, this is sitting here too long and I spent significant time on it please ;(

File: .travis.yml [hunk 5]
```diff
@@ -19,6 +19,18 @@ matrix:
...
+
+      - python: 3.8
+        env: TOX_ENV=py38-static-analysis
+
+      - python: 3.8
+        env: TOX_ENV=py38-unit-tests
+
       # For type-check, we need Python 3 to power `mypy`. It's going to check types for both
       # Python 3 *and* 2, therefore it's perfectly fine to run it just once.
       - python: 3.6
```
Comment (optak) [inline .travis.yml:36]: Why is this 3.6 part after 3.8 section and not together with other 3.6?

**Review by optak — APPROVED** [review]
LGTM

**Comment by thrix** [comment]
We have reached stable karma \o/
